mypy/newsemanal/semanal_main.py

```diff
@@ -44,22 +48,41 @@ def process_functions(graph: 'Graph', scc: List[str]) -> None:
     # TODO: This doesn't quite work yet
     # Process functions.
     deferred = []  # type: List[str]
-    for id in scc:
-        tree = graph[id].tree
+    for module in scc:
+        tree = graph[module].tree
         assert tree is not None
         symtable = tree.names
-        targets = get_all_leaf_targets(symtable)
-        for target in targets:
-            deferred += semantic_analyze_target(target, graph[id])
+        targets = get_all_leaf_targets(symtable, module, None)
+        for target, node, active_type in targets:
+            deferred += semantic_analyze_target(target, graph[module], node, active_type)
     assert not deferred  # There can't be cross-function forward refs
```

Comment:
But there can be forward references within functions, for example:
```python
def outer():
    def inner():
        print(x)
    x = 1
    inner()
```
In some sense this is similar to the `global x` problem I mentioned before.

Reply:
I think that this may be handled by another deferral mechanism inside the semantic analyzer, but I can't remember how these work exactly. Added an item about this to our task list.